compiler/pipes/final-check.cpp

```diff
@@ -41,6 +41,62 @@ void check_class_immutableness(ClassPtr klass) {
                         TermStringFormat::paint(klass->parent_class->name, TermStringFormat::red)));
 }
 
+std::list<ClassPtr> find_not_ic_compatibility_derivatives(ClassPtr klass);
+
+void check_fields_ic_compatibility(ClassPtr klass) {
+  if (klass->process_fields_ic_compatibility) {
+    return;
+  }
+  klass->process_fields_ic_compatibility = true;
+  klass->members.for_each([klass](const ClassMemberInstanceField &field) {
+    kphp_assert(field.var->marked_as_const);
+    std::unordered_set<ClassPtr> sub_classes;
+    field.var->tinf_node.get_type()->get_all_class_types_inside(sub_classes);
+    for (auto sub_class : sub_classes) {
+      // sub_class already checked for is_immutable || is_interface @see check_class_immutableness
+      check_fields_ic_compatibility(sub_class);
+      std::list<ClassPtr> descendants = find_not_ic_compatibility_derivatives(sub_class);
+      for (auto & element : descendants) {
+        kphp_error(false, fmt_format("Field {} of immutable class {} has mutable derived {}",
+                                     TermStringFormat::paint(std::string{field.local_name()}, TermStringFormat::red),
+                                     TermStringFormat::paint(klass->name, TermStringFormat::red),
+                                     TermStringFormat::paint(element->name, TermStringFormat::red)));
+      }
+    }
+  });
+  klass->process_fields_ic_compatibility = false;
+}
+
+void check_derivatives_ic_compatibility(ClassPtr klass) {
+  std::list<ClassPtr> descendants = find_not_ic_compatibility_derivatives(klass);
+  for (const auto &element : descendants) {
+    kphp_error(false, fmt_format("Can not store polymorphic type {} with mutable derived class {}", klass->name, element->name));
+  }
+}
+
+std::list<ClassPtr> find_not_ic_compatibility_derivatives(ClassPtr klass) {
+  bool has_mutable_subtree = false;
+  std::list<ClassPtr> mutable_children{};
+  for (const auto & derived : klass->derived_classes) {
+    if (derived->is_subtree_immutable == SubtreeImmutableType::immutable) {
+      // continue
+    } else if (derived->is_subtree_immutable == SubtreeImmutableType::not_immutable) {
+      // do not save them because they have already been processed
+      has_mutable_subtree = true;
+    } else if (!derived->is_immutable && !derived->is_interface()) {
+      mutable_children.push_back(derived);
+    } else {
+      check_fields_ic_compatibility(derived);
+      std::list<ClassPtr> descendant = find_not_ic_compatibility_derivatives(derived);
+      mutable_children.splice(mutable_children.end(), descendant);
+    }
```

Comment:
I suppose it could be rewritten in iterative-style DFS without extra allocations for `std::list`

Reply:
Rewrite dfs and change buffer to vector